furbysource.asm

```diff
@@ -506,4 +506,285 @@ InitIO:
+
+; Preset motor speed, assuming mid battery life, we eat the pulse width
+; so that the motor wont be runing at 6 volts and burn out. We then
+; predict what the pulse width should be for any voltage.
+
+;     LDA   #Mpulse_on  ;preset motor speed
+      LDA   #ll
+      STA   Mon_len           ;set motor on pulse timing
+
+      LDA   #05         ;
+      STA   Moff_len    ;
+
+;ÈÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍ_____ÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍ
+;* 'Diagnostics and celibration Routine                   *
+;ÈÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍ_ÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍ
+;
+
+      Include           Dieg7.asm   ;asm file
+
+; ****** Only called by diagnostic speech routines *********
+
+; Be sure to set 'MACRO_HI' and all calls are in that 128 byte block.
+
+Dieg_macro:
+      STA   Macro_Lo    ;save lo byte of Macro table entry
+        LDA     #0b8h   ;#90h            ;_ex offset to adrs_400 added
+to diag call
+      CLC
+      ADC   Macro_Lo    ;add in offset
+      STA   Macro_Lo    ;update
+      LDA   #01         ;get hi byte adrs 400 = 190h
+      STA   Macro_Hi    ;save hi byte of Macro table entry
+      JSR   Get_marcro  ;go start motor/speech
+      JSR   Notrdy            ;Do / get status for speech and motor
+      RTS               ;yo
+
+
+; Enter with Areg holding how many 30 mili second delay cycles
+
+Half_delay
+      STA   TEMP1       ;save timer
+Half d2:
+      LDA   #10         ;set 1/2 sec      (y * 2.9 mSec)
+      STA   Cycle_timer ;set it
+Half_d3:
+      LDA   Cycle_timer ;ck if done
+      BNE   Half_d3           ;loop
+      DEC   TEMP1       ;
+      BNE   Half_d2           ;loop
+      RTS               ;done
+A-21
+Test_byp:   ;We assume diagnostic only runs on coldboot
+
+;***************************************************************
+      
+      LDA   #FFh        ;initialize word training variable
+      STA   Temp_ID           ;
+
+      LDA   #FFh        ;
+      STA   Hungry_counter    ;preset furby's health
+      STA   Sick_counter
+
+;***********************************************************
+
+; We sit here and wait for tilt to go away, and just keep incrementing
+; counter until it does. This becomes the new random generator seed.
+
+Init_rnd:
+      INC   TEMP1       ;random counter
+      LDA   Port_D            ;get switches
+      AND   #03         ;check tilt & invert sw
+      BNE   Init_rnd    ;loop til gone
+      LDA   TEMP1       ;get new seed
+      STA   Spcl_seed1  ;stuff it
+      STA   Seed_1            ;also load for cold boot
+;*************************************************************
+
+; Use feed sw to generate a better random number
+
+      JSR   Get_feed    ;go test sensor
+      LDA   Stat_4            ;get system
+      AND   #Do_feed    ;ck sw
+      BNE   Feed_rnd    ;if feed sw then cold boot
+      JMP   End_coldinit      ;else do warm boot
+Feed_rnd:
+      INC   TEMP1       ;random counter
+      LDA   Stat_4            ;system
+      AND   #DFh
```

Comment:
I think this line should be:

          AND   #DFh        ;clear any prev feed sw senses